pkg/file/joiner/joiner.go

```diff
@@ -98,7 +98,7 @@ func (g *decoderCache) GetOrCreate(addrs []swarm.Address, shardCnt int) storage.
 // New creates a new Joiner. A Joiner provides Read, Seek and Size functionalities.
 func New(ctx context.Context, g storage.Getter, putter storage.Putter, address swarm.Address) (file.Joiner, int64, error) {
 	// retrieve the root chunk to read the total data length the be retrieved
-	rLevel := replicas.GetLevelFromContext(ctx)
+	rLevel := redundancy.GetLevelFromContext(ctx)
```

Comment:
Also, the API doc says default is supposed to be NONE. Not PARANOID. Either the API doc needs updating or the default value needs to be updated.

Reply:
First of all the default in the API doc talks about NONE being the default in the context of the encoding only.
As for the erasure coding, level is encoded in the span, so the level of redundancy is known by the decoder at download.

But in the case of replicas, the level is not known by the downloader. Therefore we choose to fall back to trying all replicas as if the uploader had been PARANOID.
The way replicas work for retrieval is that initially only the root chunk is retrieved, then we fall back to extending it to replicas which are present assuming the next level of security for encoding. Indeed in the definition the replicas for level 3 are a superset of level 3 and a subset of level 4.

Note that even if the encoding is 'known', it would mean we know that the publisher intended to upload say 4 replicas (level 2). However, by design replicas can be added by third parties, therefore you might as well try just in case.

Also, note that attempting replicas belonging only to higher levels of redundancy are attempted only when desparate, ie., if lower levels are not found.

A user configurablee header value should probably be introduced and set on the context so that the user can control the swap traffic.

Reply:
> A user configurable header value should probably be introduced and set on the context so that the user can control the swap traffic.

My vote would go for that!  Especially when each node of a /bzz manifest retrieval would be using replicas as many of them fit easily within a single 4K chunk.